lms/services/h_api.py

```diff
@@ -0,0 +1,180 @@
+            data={"display_name": h_user.display_name},
+        )
+
+    def upsert_user(self, h_user, provider, provider_unique_id):
+        """
+        Create or update a user in H as appropriate.
+
+        This is equivalent to calling `update_user()` then `create_user()` if
+        that fails.
+
+        :param h_user: An `HUser` object to send to H
+        :param provider: The provider (hosting LMS)
+        :param provider_unique_id: The id of the provider
+        """
+        try:
+            self.update_user(h_user)
+        except HAPINotFoundError:
+            self.create_user(h_user, provider, provider_unique_id)
+
+    def create_group(self, group_id, group_name, h_user):
+        """
+        Create a group in H.
+
+        :param group_id: The id of the group
+        :param group_name: The display name for the group
+        :param h_user: An `HUser` to associate with the group
+        """
+        self._api_request(
+            "PUT",
+            f"groups/{group_id}",
+            data={"groupid": group_id, "name": group_name},
+            headers={"X-Forwarded-User": h_user.userid},
+        )
+
+    def update_group(self, group_id, group_name):
+        """
+        Update a group in H.
+
+        Currently this only updates the name of the group.
+
+        :param group_id: The id of the group
+        :param group_name: The display name for the group
+        """
+        self._api_request(
+            "PATCH", f"groups/{group_id}", data={"name": group_name},
+        )
+
+    def add_user_to_group(self, h_user, group_id):
+        """
+        Add the user as a member of the group.
+
+        :param h_user: An `HUser` to associate with the group
+        :param group_id: The id of the group
+        """
+        self._api_request("POST", f"groups/{group_id}/members/{h_user.userid}")
+
+    def _api_request(
+        self, method: str, path: str, data: dict = None, headers: dict = None
+    ):
+        """
+        Send any kind of HTTP request to the h API and return the response.
+
+        :param method: the HTTP request method to use, for example "GET",
+                       "POST", "PUT", "PATCH" or "DELETE"
+        :param path: the h API path to post to, relative to
+                     ``settings["h_api_url_private"]``, for example:
+                     ``"users"`` or ``"groups/<GROUPID>/members/<USERID>"``
+        :param data: the data to post as JSON in the request body
+        :param headers: extra headers to pass with the request
+
+        :raise HAPINotFoundError: If the request fails with 404
+        :raise HAPIError: if the request fails for any other reason
+        :return: the response from the h API
+        :rtype: requests.Response
+        """
+        headers = headers or {}
+        headers["Hypothesis-Application"] = "lms"
+
+        request_args = {"headers": headers}
+
+        if data is not None:
+            request_args["data"] = json.dumps(data, separators=(",", ":"))
```

Comment:
The separators here compactify the JSON. Not much but not really any point in wasting the space in the request